File: Leetcode/Python3/641. Design Circular Deque/solution.py
# https://leetcode.com/problems/design-circular-deque/


class MyCircularDeque:
    def __init__(self, k: int):
        self._deque = [None for _ in range(k)]
        self._size = 0
        self._maxlen = k
        self._front = 0
        self._rear = 0

    def insertFront(self, value: int) -> bool:
        if self.isFull():
            return False
        elif self.isEmpty():
            self._deque[self._front] = value
        else:
            self._front = (self._front - 1) % self._maxlen
            self._deque[self._front] = value
        self._size += 1
        return True

    def insertLast(self, value: int) -> bool:
        if self.isFull():
            return False
        elif self.isEmpty():
            self._deque[self._rear] = value
        else:
            self._rear = (self._rear + 1) % self._maxlen
            self._deque[self._rear] = value
        self._size += 1
        return True

    def deleteFront(self) -> bool:
        if self.isEmpty():
            return False
        self._deque[self._front] = None
        self._front = (self._front + 1) % self._maxlen
        self._size -= 1
        if self.isEmpty():  # When self._size == 1 front, rear pointer dont match after delete
            self._front = self._rear
        return True

    def deleteLast(self) -> bool:
        if self.isEmpty():
            return False
        self._deque[self._rear] = None
        self._rear = (self._rear - 1) % self._maxlen
        self._size -= 1
        if self.isEmpty():  # When self._size == 1 front, rear pointer dont match after delete
            self._front = self._rear
        return True

    def getFront(self) -> int:
        if self.isEmpty():
            return -1
        return self._deque[self._front]

    def getRear(self) -> int:
        if self.isEmpty():
            return -1
        return self._deque[self._rear]

    def isEmpty(self) -> bool:
        return self._size == 0

    def isFull(self) -> bool:
        return self._size == self._maxlen


# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()

File: Leetcode/C++/641. Design Circular Deque/solution.cpp
// https://leetcode.com/problems/design-circular-deque/
// Ref: https://leetcode.com/problems/design-circular-deque/solutions/155209/c-99-ring-buffer-no-edge-cases-fb-interviewer-really-loves-it-easy-to-impl-in-4mins-cheers/

#include <vector>

using namespace std;

class MyCircularDeque
{
public:
    vector<int> v;
    int maxSize, front, rear = 0, curSize = 0;
    MyCircularDeque(int k)
    {
        front = k - 1;
        v.resize(k);
        maxSize = k;
    }

    bool insertFront(int value)
    {
        if (isFull())
        {
            return false;
        }
        v[front] = value;
        if (front == 0)
        {
            front = maxSize - 1;
        }
        else
        {
            front -= 1;
        }
        curSize += 1;
        return true;
    }

    bool insertLast(int value)
    {
        if (isFull())
        {
            return false;
        }
        v[rear] = value;
        rear = (rear + 1) % maxSize;
        curSize += 1;
        return true;
    }

    bool deleteFront()
    {
        if (isEmpty())
        {
            return false;
        }
        front = (front + 1) % maxSize;
        curSize -= 1;
        return true;
    }

    bool deleteLast()
    {
        if (isEmpty())
        {
            return false;
        }
        if (rear == 0)
        {
            rear = maxSize - 1;
        }
        else
        {
            rear -= 1;
        }
        curSize -= 1;
        return true;
    }

    int getFront()
    {
        return isEmpty() ? -1 : v[(front + 1) % maxSize];
    }

    int getRear()
    {
        if (isEmpty())
        {
            return -1;
        }
        if (rear == 0)
        {
            return v[maxSize - 1];
        }
        return v[rear - 1];
    }

    bool isEmpty()
    {
        return curSize == 0;
    }

    bool isFull()
    {
        return curSize == maxSize;
    }
};

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque* obj = new MyCircularDeque(k);
 * bool param_1 = obj->insertFront(value);
 * bool param_2 = obj->insertLast(value);
 * bool param_3 = obj->deleteFront();
 * bool param_4 = obj->deleteLast();
 * int param_5 = obj->getFront();
 * int param_6 = obj->getRear();
 * bool param_7 = obj->isEmpty();
 * bool param_8 = obj->isFull();
 */